src/syntax/FStar.Syntax.Util.fs

```diff
@@ -1362,37 +1362,56 @@ type connective =
     | QEx of binders * qpats * typ
     | BaseConn of lident * args
 
+(* destruct_typ_as_formula can be a bottleneck
+   these tables are defined here to ensure they are constructed
+   only once in the compiled extracted code
+
+   the tables encode arity -> lids
+   and in the case of sq_base the lid mapping
+   *)
+let destruct_base_table =
+  [ (0, [PC.true_lid; PC.false_lid]);
+    (2, [PC.and_lid; PC.or_lid; PC.imp_lid; PC.iff_lid; PC.eq2_lid; PC.eq3_lid]);
+    (1, [PC.not_lid]);
+    (3, [PC.ite_lid; PC.eq2_lid]);
+    (4, [PC.eq3_lid])
+  ]
+
+let destruct_sq_base_table =
+  [ (2, [(PC.c_and_lid, PC.and_lid); (PC.c_or_lid, PC.or_lid);
+         (PC.c_eq2_lid, PC.c_eq2_lid); (PC.c_eq3_lid, PC.c_eq3_lid)]);
+    (3, [(PC.c_eq2_lid, PC.c_eq2_lid)]);
+    (4, [(PC.c_eq3_lid, PC.c_eq3_lid)]);
+    (0, [(PC.c_true_lid, PC.true_lid); (PC.c_false_lid, PC.false_lid)])
+  ]
+
 let destruct_typ_as_formula f : option<connective> =
     let rec unmeta_monadic f =
       let f = Subst.compress f in
       match f.n with
       | Tm_meta(t, Meta_monadic _)
       | Tm_meta(t, Meta_monadic_lift _) -> unmeta_monadic t
       | _ -> f in
-    let destruct_base_conn f =
-        let connectives = [ (PC.true_lid,  0);
-                            (PC.false_lid, 0);
-                            (PC.and_lid,   2);
-                            (PC.or_lid,    2);
-                            (PC.imp_lid, 2);
-                            (PC.iff_lid, 2);
-                            (PC.ite_lid, 3);
-                            (PC.not_lid, 1);
-                            (PC.eq2_lid, 3);
-                            (PC.eq2_lid, 2);
-                            (PC.eq3_lid, 4);
-                            (PC.eq3_lid, 2)
-                        ]
-        in
+    let mk_base_conn lid args = Some (BaseConn (lid, args)) in
 
-        let aux f (lid, arity) =
-            let t, args = head_and_args (unmeta_monadic f) in
-            let t = un_uinst t in
-            if is_constructor t lid
-            && List.length args = arity
-            then Some (BaseConn(lid, args))
-            else None in
-        U.find_map connectives (aux f) in
+    let destruct_base_conn f =
+        let t, args = head_and_args (unmeta_monadic f) in
+        let t = un_uinst t in
+        match (pre_typ t).n with
```

Comment:
I'd like to get a set of expert eyes here on the usage of `pre_typ` and why it's needed. I'm not familiar with this helper function, and don't know whether e.g. `compress` would be more suitable.

@nikswamy @mtzguido

Reply:
Yeah, this looks odd and different to what `destruct_sq_base_conn` is doing, I'm trying out a CI build just using `compress`.

Reply:
Got a green in `guido_test1827`, I would add the changes there to this PR and merge.